libsrc/leddevice/dev_net/LedDeviceYeelight.cpp

```diff
@@ -0,0 +1,1503 @@
+
+	if ( !apiHostname.isEmpty() )
+	{
+		YeelightLight yeelight(_log, apiHostname, apiPort);
+
+		//yeelight.setDebuglevel(3);
+		if ( yeelight.open() )
+		{
+			properties.insert("properties", yeelight.getProperties());
+			yeelight.close();
+		}
+	}
+	Debug(_log, "properties: [%s]", QString(QJsonDocument(properties).toJson(QJsonDocument::Compact)).toUtf8().constData() );
+
+	return properties;
+}
+
+void LedDeviceYeelight::identify(const QJsonObject& params)
+{
+	Debug(_log, "params: [%s]", QString(QJsonDocument(params).toJson(QJsonDocument::Compact)).toUtf8().constData() );
+
+	QString apiHostname = params["hostname"].toString("");
+	quint16 apiPort = static_cast<quint16>( params["port"].toInt(API_DEFAULT_PORT) );
+	Debug (_log, "apiHost [%s], apiPort [%d]", QSTRING_CSTR(apiHostname), apiPort);
+
+	if ( !apiHostname.isEmpty() )
+	{
+		YeelightLight yeelight(_log, apiHostname, apiPort);
+		//yeelight.setDebuglevel(3);
+
+		if ( yeelight.open() )
+		{
+			yeelight.identify();
+			yeelight.close();
+		}
+	}
+}
+
+int LedDeviceYeelight::write(const std::vector<ColorRgb> & ledValues)
+{
+	//DebugIf(verbose, _log, "enabled [%d], _isDeviceReady [%d]", _isEnabled, _isDeviceReady);
+	int rc = -1;
+
+	//Update on all Yeelights by iterating through lights and set colors.
+	unsigned int idx = 0;
+	int lightsInError = 0;
+	for (YeelightLight& light : _lights)
+	{
+		// Get color
+		ColorRgb color = ledValues.at(idx);
+
+		if ( light.isReady()  )
+		{
+			bool skipWrite = false;
+			if ( !light.isInMusicMode() )
+			{
+				if ( light.setMusicMode(true, _musicModeServerAddress, _musicModeServerPort) )
+				{
+					// Wait for callback of the device to establish streaming socket
+					if ( _tcpMusicModeServer->waitForNewConnection(CONNECT_STREAM_TIMEOUT) )
+					{
+						light.setStreamSocket( _tcpMusicModeServer->nextPendingConnection() );
+					}
+					else
+					{
+						QString errorReason = QString ("(%1) %2").arg(_tcpMusicModeServer->serverError()).arg( _tcpMusicModeServer->errorString());
+						Warning( _log, "write Error [%s]: _tcpMusicModeServer: %s", QSTRING_CSTR(light.getName()), QSTRING_CSTR(errorReason));
+						light.setInError("Failed to get stream socket");
+					}
+				}
+				else
+				{
+					DebugIf(verbose,_log, "setMusicMode failed due to command quota issue, skip write and try with next");
+					skipWrite = true;
+				}
+			}
+
+			if ( !skipWrite )
+			{
+				// Update light with given color
+				if ( _outputColorModel == MODEL_RGB )
+				{
+					light.setColorRGB( color );
+				}
+				else
+				{
+					light.setColorHSV( color );
+				}
+			}
+		}
+		else
+		{
+			++lightsInError;
+		}
+		++idx;
+	}
+
+	if ( ! (lightsInError < static_cast<int>(_lights.size())) )
+	{
+		this->setInError( "All Yeelights in error - stopping device!" );
+	}
+	else
+	{
+		// Minimum one Yeelight device is working, continue updating devices
+		rc = 0;
+	}
+
+	//DebugIf(verbose, _log, "rc [%d]", rc );
+
+	return rc;
+}
```

Comment:
1500 lines for a single CPP hints that some functionality needs to be moved elsewhere.

Reply:
Fair point. I may need to have the individual YeelightLight class separate from the YeelightDevice... (then I could move ~1000 lines to a separate ccp. Need the to think about filenaming and where to put those sub-object classes. PhilipsHueLights would be the 2nd candidate...

Plus change my readability style....